File: examples/php/ejemplo_basico.php
<?php
/**
 * Ejemplo básico de uso de JVDataAccess en PHP
 * Versión: 1.0.0
 * 
 * Este ejemplo demuestra las funcionalidades básicas de:
 * - YourSQL: Conector MySQL personalizado
 * - JVDB: Clase de abstracción para base de datos
 */

// Cargar configuración
require_once __DIR__ . '/../../config.php';

// Cargar clases
require_once __DIR__ . '/../../src/php/JVDB/JVDB.php';

use JVDB\JVDB;

function imprimirSeparador($titulo) {
    echo "\n" . str_repeat("=", 80) . "\n";
    echo "📚 $titulo\n";
    echo str_repeat("=", 80) . "\n";
}

function imprimirSubtitulo($numero, $titulo) {
    echo "\n$numero  $titulo:\n";
    echo str_repeat("-", 60) . "\n";
}

/**
 * Ejemplo básico de uso de JVDB
 */
function ejemploJVDBBasico() {
    imprimirSeparador("EJEMPLO 1: Uso básico de JVDB");
    
    // Crear instancia de JVDB (sin echo de conexión para este ejemplo)
    ob_start();
    $db = new JVDB(DB_HOST, DB_USER, DB_PASSWORD, DB_NAME);
    ob_end_clean();
    
    // 1. Listar todas las tablas
    imprimirSubtitulo("1️⃣", "Listar tablas de la base de datos");
    $tablas = $db->tablas('array');
    foreach ($tablas as $tabla) {
        echo "  📋 $tabla\n";
    }
    
    // 2. Consultar productos
    imprimirSubtitulo("2️⃣", "Consultar todos los productos");
    $productos = $db->seleccionar('productos', '*', 'array');
    foreach (array_slice($productos, 0, 3) as $producto) {
        echo "  🛍️  {$producto['nombre']} - \${$producto['precio']}\n";
    }
    
    // 3. Buscar productos por categoría
    imprimirSubtitulo("3️⃣", "Buscar productos de categoría 'Periféricos'");
    $perifericos = $db->buscar('productos', 'categoria', 'Periféricos', 'array');
    foreach ($perifericos as $producto) {
        echo "  ⌨️  {$producto['nombre']} - Stock: {$producto['stock']}\n";
    }
    
    // 4. Insertar un nuevo producto
    imprimirSubtitulo("4️⃣", "Insertar nuevo producto");
    $nuevoProducto = [
        'nombre' => 'Micrófono Blue Yeti',
        'descripcion' => 'Micrófono USB profesional',
        'precio' => 129.99,
        'stock' => 15,
        'categoria' => 'Audio'
    ];
    $db->insertar('productos', $nuevoProducto);
    
    // 5. Consultar un registro específico
    imprimirSubtitulo("5️⃣", "Consultar cliente por ID");
    $cliente = $db->seleccionarUno('clientes', 1);
    if ($cliente) {
        echo "  👤 {$cliente['nombre']} {$cliente['apellidos']}\n";
        echo "  📧 {$cliente['email']}\n";
    }
    
    // Cerrar conexión
    unset($db);
    echo "\n✅ Ejemplo completado con éxito\n";
}

/**
 * Ejemplo de consultas más avanzadas con JVDB
 */
function ejemploConsultasAvanzadas() {
    imprimirSeparador("EJEMPLO 2: Consultas avanzadas con JVDB");
    
    ob_start();
    $db = new JVDB(DB_HOST, DB_USER, DB_PASSWORD, DB_NAME);
    ob_end_clean();
    
    // 1. Consulta personalizada con JOIN
    imprimirSubtitulo("1️⃣", "Consulta con JOIN - Pedidos con datos de clientes");
    $query = "
        SELECT 
            p.Identificador,
            c.nombre,
            c.email,
            p.fecha_pedido,
            p.estado,
            p.total
        FROM pedidos p
        INNER JOIN clientes c ON p.cliente_id = c.Identificador
        ORDER BY p.fecha_pedido DESC
        LIMIT 5
    ";
    $pedidos = $db->consultaPersonalizada($query, null, 'array');
    foreach ($pedidos as $pedido) {
        echo "  📦 Pedido #{$pedido['Identificador']} - {$pedido['nombre']}\n";
        echo "     Estado: {$pedido['estado']} | Total: \${$pedido['total']}\n";
    }
    
    // 2. Actualizar un producto
    imprimirSubtitulo("2️⃣", "Actualizar stock de un producto");
    $db->actualizar('productos', 2, ['stock' => 60]);
    $producto = $db->seleccionarUno('productos', 2);
    echo "  Stock actualizado: {$producto['nombre']} -> {$producto['stock']} unidades\n";
    
    // 3. Estructura de una tabla
    imprimirSubtitulo("3️⃣", "Ver estructura de tabla 'productos'");
    $estructura = $db->estructuraTabla('productos', 'array');
    foreach ($estructura as $columna) {
        echo "  🔹 {$columna['Field']} - {$columna['Type']}\n";
    }
    
    // 4. Uso de transacciones
    imprimirSubtitulo("4️⃣", "Ejemplo de transacción");
    try {
        $db->iniciarTransaccion();
        
        // Insertar un nuevo cliente
        $nuevoCliente = [
            'nombre' => 'Luis',
            'apellidos' => 'González Martín',
            'email' => 'luis.gonzalez@email.com',
            'telefono' => '600678901',
            'ciudad' => 'Sevilla'
        ];
        $db->insertar('clientes', $nuevoCliente);
        
        // Confirmar transacción
        $db->confirmarTransaccion();
        
    } catch (Exception $e) {
        echo "  ❌ Error: " . $e->getMessage() . "\n";
        $db->revertirTransaccion();
    }
    
    unset($db);
    echo "\n✅ Ejemplo completado con éxito\n";
}

/**
 * Ejemplo de uso directo de YourSQL
 */
function ejemploYourSQLDirecto() {
    imprimirSeparador("EJEMPLO 3: Uso directo de YourSQL (bajo nivel)");
    
    require_once __DIR__ . '/../../src/php/YourSQL/YourSQL.php';
    
    // Crear conexión con YourSQL (usando namespace completo)
    $conexion = new \YourSQL\YourSQLConnection(DB_HOST, DB_USER, DB_PASSWORD, DB_NAME);
    $conexion->connect();
    echo "\n✅ Conexión establecida con YourSQL\n";
    
    // 1. Query directo
    imprimirSubtitulo("1️⃣", "Consulta directa con YourSQL");
    $resultado = $conexion->executeQuery("SELECT * FROM productos WHERE precio < 100");
    foreach ($resultado as $producto) {
        echo "  💰 {$producto['nombre']} - \${$producto['precio']}\n";
    }
    
    // 2. Query con parámetros (prepared statement)
    imprimirSubtitulo("2️⃣", "Consulta preparada con parámetros");
    $resultado = $conexion->executeQuery(
        "SELECT * FROM productos WHERE categoria = ?",
        ['Audio']
    );
    foreach ($resultado as $producto) {
        echo "  🎧 {$producto['nombre']}\n";
    }
    
    // 3. Query Builder
    imprimirSubtitulo("3️⃣", "Uso del Query Builder");
    $builder = new \YourSQL\YourSQLQueryBuilder($conexion);
    $resultado = $builder
        ->select('nombre', 'email', 'ciudad')
        ->from('clientes')
        ->where("ciudad = 'Barcelona'")
        ->orderBy('nombre', 'ASC')
        ->execute();
    
    foreach ($resultado as $cliente) {
        echo "  🌆 {$cliente['nombre']} - {$cliente['ciudad']}\n";
    }
    
    // 4. Información de tablas
    imprimirSubtitulo("4️⃣", "Listar todas las tablas");
    $tablas = $conexion->getTables();
    foreach ($tablas as $tabla) {
        echo "  📋 $tabla\n";
    }
    
    $conexion->disconnect();
    echo "\n✅ Ejemplo completado con éxito\n";
}

/**
 * Menú interactivo
 */
function menuPrincipal() {
    while (true) {
        echo "\n" . str_repeat("=", 80) . "\n";
        echo "🚀 JVDataAccess - Ejemplos de Uso (PHP)\n";
        echo str_repeat("=", 80) . "\n";
        echo "\nSelecciona un ejemplo:\n";
        echo "  1. Uso básico de JVDB\n";
        echo "  2. Consultas avanzadas\n";
        echo "  3. Uso directo de YourSQL\n";
        echo "  4. Ejecutar todos los ejemplos\n";
        echo "  0. Salir\n";
        echo str_repeat("-", 80) . "\n";
        
        echo "\n👉 Opción: ";
        $opcion = trim(fgets(STDIN));
        
        switch ($opcion) {
            case '1':
                ejemploJVDBBasico();
                break;
            case '2':
                ejemploConsultasAvanzadas();
                break;
            case '3':
                ejemploYourSQLDirecto();
                break;
            case '4':
                ejemploJVDBBasico();
                ejemploConsultasAvanzadas();
                ejemploYourSQLDirecto();
                break;
            case '0':
                echo "\n👋 ¡Hasta luego!\n\n";
                exit(0);
            default:
                echo "\n❌ Opción no válida\n";
        }
        
        echo "\nPresiona ENTER para continuar...";
        fgets(STDIN);
    }
}

// Punto de entrada
echo "\n⚠️  Asegúrate de haber ejecutado el script database/init.sql primero\n";
echo "    y de configurar tus credenciales en config.php\n\n";
echo "¿Continuar? (s/n): ";
$respuesta = trim(fgets(STDIN));

if (strtolower($respuesta) === 's') {
    menuPrincipal();
} else {
    echo "\n👋 ¡Hasta luego!\n\n";
}

File: examples/python/ejemplo_basico.py
"""
Ejemplo básico de uso de JVDataAccess en Python
Versión: 1.0.0

Este ejemplo demuestra las funcionalidades básicas de:
- YourSQL: Conector MySQL personalizado
- JVDB: Clase de abstracción para base de datos
"""

import sys
import os

# Añadir el directorio src al path para importar los módulos
sys.path.insert(0, os.path.join(os.path.dirname(__file__), '..', '..', 'src', 'python'))

from jvdb import JVDB
from yoursql import YourSQLConnection, YourSQLQueryBuilder
import json

# Importar configuración
sys.path.insert(0, os.path.join(os.path.dirname(__file__), '..', '..'))
from config import DB_CONFIG


def ejemplo_jvdb_basico():
    """Ejemplo básico de uso de JVDB"""
    print("=" * 80)
    print("📚 EJEMPLO 1: Uso básico de JVDB")
    print("=" * 80)
    
    # Crear instancia de JVDB
    db = JVDB(
        host=DB_CONFIG['host'],
        usuario=DB_CONFIG['user'],
        contrasena=DB_CONFIG['password'],
        basedatos=DB_CONFIG['database']
    )
    
    # 1. Listar todas las tablas
    print("\n1️⃣  Listar tablas de la base de datos:")
    print("-" * 40)
    tablas = db.tablas(formato='list')
    for tabla in tablas:
        print(f"  📋 {tabla}")
    
    # 2. Consultar productos
    print("\n2️⃣  Consultar todos los productos:")
    print("-" * 40)
    productos = db.seleccionar('productos', formato='list')
    for producto in productos[:3]:  # Mostrar solo 3
        print(f"  🛍️  {producto['nombre']} - ${producto['precio']}")
    
    # 3. Buscar productos por categoría
    print("\n3️⃣  Buscar productos de categoría 'Periféricos':")
    print("-" * 40)
    perifericos = db.buscar('productos', 'categoria', 'Periféricos', formato='list')
    for producto in perifericos:
        print(f"  ⌨️  {producto['nombre']} - Stock: {producto['stock']}")
    
    # 4. Insertar un nuevo producto
    print("\n4️⃣  Insertar nuevo producto:")
    print("-" * 40)
    nuevo_producto = {
        'nombre': 'Webcam Logitech C920',
        'descripcion': 'Cámara web Full HD 1080p',
        'precio': 79.99,
        'stock': 25,
        'categoria': 'Periféricos'
    }
    db.insertar('productos', nuevo_producto)
    
    # 5. Consultar un registro específico
    print("\n5️⃣  Consultar cliente por ID:")
    print("-" * 40)
    cliente = db.seleccionar_uno('clientes', 1)
    if cliente:
        print(f"  👤 {cliente['nombre']} {cliente['apellidos']}")
        print(f"  📧 {cliente['email']}")
    
    # Cerrar conexión
    db.desconectar()
    print("\n✅ Ejemplo completado con éxito\n")


def ejemplo_consultas_avanzadas():
    """Ejemplo de consultas más avanzadas con JVDB"""
    print("=" * 80)
    print("📚 EJEMPLO 2: Consultas avanzadas con JVDB")
    print("=" * 80)
    
    db = JVDB(
        host=DB_CONFIG['host'],
        usuario=DB_CONFIG['user'],
        contrasena=DB_CONFIG['password'],
        basedatos=DB_CONFIG['database']
    )
    
    # 1. Consulta personalizada con JOIN
    print("\n1️⃣  Consulta con JOIN - Pedidos con datos de clientes:")
    print("-" * 60)
    query = """
        SELECT 
            p.Identificador,
            c.nombre,
            c.email,
            p.fecha_pedido,
            p.estado,
            p.total
        FROM pedidos p
        INNER JOIN clientes c ON p.cliente_id = c.Identificador
        ORDER BY p.fecha_pedido DESC
        LIMIT 5
    """
    pedidos = db.consulta_personalizada(query, formato='list')
    for pedido in pedidos:
        print(f"  📦 Pedido #{pedido['Identificador']} - {pedido['nombre']}")
        print(f"     Estado: {pedido['estado']} | Total: ${pedido['total']}")
    
    # 2. Actualizar un producto
    print("\n2️⃣  Actualizar stock de un producto:")
    print("-" * 60)
    db.actualizar('productos', 2, {'stock': 55})
    producto = db.seleccionar_uno('productos', 2)
    print(f"  Stock actualizado: {producto['nombre']} -> {producto['stock']} unidades")
    
    # 3. Estructura de una tabla
    print("\n3️⃣  Ver estructura de tabla 'productos':")
    print("-" * 60)
    estructura = db.estructura_tabla('productos', formato='list')
    for columna in estructura:
        print(f"  🔹 {columna['Field']} - {columna['Type']}")
    
    db.desconectar()
    print("\n✅ Ejemplo completado con éxito\n")


def ejemplo_yoursql_directo():
    """Ejemplo de uso directo de YourSQL"""
    print("=" * 80)
    print("📚 EJEMPLO 3: Uso directo de YourSQL (bajo nivel)")
    print("=" * 80)
    
    # Crear conexión con YourSQL
    conexion = YourSQLConnection(
        host=DB_CONFIG['host'],
        user=DB_CONFIG['user'],
        password=DB_CONFIG['password'],
        database=DB_CONFIG['database']
    )
    
    conexion.connect()
    print("\n✅ Conexión establecida con YourSQL\n")
    
    # 1. Query directo
    print("1️⃣  Consulta directa con YourSQL:")
    print("-" * 60)
    resultado = conexion.execute_query("SELECT * FROM productos WHERE precio < 100")
    for producto in resultado:
        print(f"  💰 {producto['nombre']} - ${producto['precio']}")
    
    # 2. Query con parámetros (prepared statement)
    print("\n2️⃣  Consulta preparada con parámetros:")
    print("-" * 60)
    resultado = conexion.execute_query(
        "SELECT * FROM productos WHERE categoria = %s",
        params=('Audio',)
    )
    for producto in resultado:
        print(f"  🎧 {producto['nombre']}")
    
    # 3. Query Builder
    print("\n3️⃣  Uso del Query Builder:")
    print("-" * 60)
    builder = YourSQLQueryBuilder(conexion)
    resultado = (builder
                .select('nombre', 'email', 'ciudad')
                .from_table('clientes')
                .where("ciudad = 'Barcelona'")
                .order_by('nombre', 'ASC')
                .execute())
    
    for cliente in resultado:
        print(f"  🌆 {cliente['nombre']} - {cliente['ciudad']}")
    
    conexion.disconnect()
    print("\n✅ Ejemplo completado con éxito\n")


def ejemplo_context_manager():
    """Ejemplo usando context managers (with statement)"""
    print("=" * 80)
    print("📚 EJEMPLO 4: Uso de Context Managers")
    print("=" * 80)
    
    print("\n💡 Los context managers gestionan automáticamente la conexión\n")
    
    # Usar JVDB con context manager
    with JVDB(
        host=DB_CONFIG['host'],
        usuario=DB_CONFIG['user'],
        contrasena=DB_CONFIG['password'],
        basedatos=DB_CONFIG['database'],
        autoconnect=True
    ) as db:
        print("Dentro del contexto - conexión activa")
        productos = db.seleccionar('productos', columnas='nombre, precio', formato='list')
        print(f"Se obtuvieron {len(productos)} productos")
    
    print("Fuera del contexto - conexión cerrada automáticamente\n")
    print("✅ Ejemplo completado con éxito\n")


def menu_principal():
    """Menú interactivo para ejecutar los ejemplos"""
    while True:
        print("\n" + "=" * 80)
        print("🚀 JVDataAccess - Ejemplos de Uso (Python)")
        print("=" * 80)
        print("\nSelecciona un ejemplo:")
        print("  1. Uso básico de JVDB")
        print("  2. Consultas avanzadas")
        print("  3. Uso directo de YourSQL")
        print("  4. Context Managers")
        print("  5. Ejecutar todos los ejemplos")
        print("  0. Salir")
        print("-" * 80)
        
        opcion = input("\n👉 Opción: ").strip()
        
        if opcion == '1':
            ejemplo_jvdb_basico()
        elif opcion == '2':
            ejemplo_consultas_avanzadas()
        elif opcion == '3':
            ejemplo_yoursql_directo()
        elif opcion == '4':
            ejemplo_context_manager()
        elif opcion == '5':
            ejemplo_jvdb_basico()
            ejemplo_consultas_avanzadas()
            ejemplo_yoursql_directo()
            ejemplo_context_manager()
        elif opcion == '0':
            print("\n👋 ¡Hasta luego!\n")
            break
        else:
            print("\n❌ Opción no válida\n")
        
        input("\nPresiona ENTER para continuar...")


if __name__ == "__main__":
    # Verificar que la base de datos esté configurada
    print("\n⚠️  Asegúrate de haber ejecutado el script database/init.sql primero")
    print("    y de configurar tus credenciales en config.py\n")
    
    respuesta = input("¿Continuar? (s/n): ").strip().lower()
    if respuesta == 's':
        menu_principal()
    else:
        print("\n👋 ¡Hasta luego!\n")
